skbio/diversity/beta/_fast_unifrac.py

```diff
@@ -0,0 +1,444 @@
+    ----------
+    branch_lengths: np.array
+         branch_lengths should be row vector, same length as # nodes in tree.
+    i : np.array
+        a slice of states from m, same length as # nodes in tree.
+        Slicing m (e.g. m[:,i]) returns a vector in the right format; note
+        that it should be a row vector (the default), not a column vector.
+    j : np.array
+        a slice of states from m, same length as # nodes in tree.
+        Slicing m (e.g. m[:,j]) returns a vector in the right format; note
+        that it should be a row vector (the default), not a column vector.
+
+    Returns
+    -------
+    float
+        Unweighted unifrac metric
+
+    Notes
+    -----
+    This is cogent.maths.unifrac.fast_tree.unifrac, but there are
+    other metrics that can (should?) be ported like:
+     - unnormalized_unifrac
+     - G
+     - unnormalized_G
+    """
+
+    _or, _and = np.logical_or(i, j), np.logical_and(i, j)
+    return 1 - ((branch_lengths * _and).sum() / (branch_lengths * _or).sum())
+
+
+def w_unifrac(branch_lengths, tip_indices, i, j):
+    """Calculates weighted unifrac(i,j) from branch lengths and cols i,j of m.
+    """
+    i_sum = (np.take(i, tip_indices)).sum()
+    j_sum = (np.take(j, tip_indices)).sum()
+
+    if i_sum:
+        i_ = i / i_sum
+    else:
+        i_ = 0.0
+
+    if j_sum:
+        j_ = j / j_sum
+    else:
+        j_ = 0.0
+
+    return (branch_lengths * abs(i_ - j_)).sum()
+
+
+def _skbio_counts_to_envs(otu_ids, *counts):
+    """This is a place holder method to help get the API hooked up
+
+    This method should be unnecessary once the implementation is optimized to
+    the interface.
+    """
+    envs = {}
+    n_counts = len(counts)
+
+    for packed in zip(otu_ids, *counts):
+        # NOTE: this is ducttape to fit the API
+        otu_id = packed[0]
+
+        counts = {}
+        for env in range(1, n_counts + 1):
+            if packed[env]:
+                counts[env] = packed[env]
+
+        if counts:
+            envs[otu_id] = counts
+
+    return envs
+
+
+def unweighted_unifrac(u_counts, v_counts, otu_ids, tree,
+                       suppress_validation=False):
+    """fit to unifrac API"""
+    if not suppress_validation:
+        _validate(u_counts, v_counts, otu_ids, tree)
+
+    u_sum = sum(u_counts)
+    v_sum = sum(v_counts)
+
+    if not u_sum or not v_sum:
+        if u_sum + v_sum:
+            # u or v counts are all zeros
+            return 1.0
+        else:
+            # u and v are zero
+            return 0.0
+
+    envs = _skbio_counts_to_envs(otu_ids, u_counts, v_counts)
+
+    return fast_unifrac(tree, envs)
+
+
+def weighted_unifrac(u_counts, v_counts, otu_ids, tree, normalized=False,
+                     suppress_validation=False):
+    u_sum = sum(u_counts)
+    v_sum = sum(v_counts)
+
+    if not u_sum or not v_sum:
+        if u_sum + v_sum:
+            # u or v counts are all zeros
+            if normalized:
+                return 1.0
```

Comment:
Can this fall through to my implementation for the time-being? Or raise a ``NotImplementedError``?